### PR TITLE
Expand resource copying APIs

### LINKED ISSUE #158: Improve and streamline resource copying APIs
### Description

This proposal is to do a refactor/revamp of the resource copying APIs. This is to fix these issues:

- Some inconsistencies
- Some missing features (eg. copying between 2D textures)
- Some missing mirrored APIs (eg. `CopyTo` but no `CopyFrom`)

### Rationale

Copying between resources should be as easy to do as possible, as it's a very common operations.
Not having the necessary APIs immediately available risks pushing developers to use slower patterns like temporary arrays.

### Proposed API

cc. @hawkerm @Avid29 @Patan77 @ahmed605 for feedbacks in case I missed anything in the API breakdown below.

```diff
 public abstract class Buffer<T>
 {
-    public abstract void CopyFrom(Buffer<T> source);
 }

 public static class BufferExtensions
 {
     public static T[] ToArray<T>(this Buffer<T> buffer) where T : unmanaged;
     public static T[] ToArray<T>(this Buffer<T> buffer, int offset, int count) where T : unmanaged;
     public static void CopyTo<T>(this Buffer<T> buffer, T[] destination) where T : unmanaged;
-    public static void CopyTo<T>(this Buffer<T> buffer, T[] destination, int destinationOffset, int bufferOffset, int count) where T : unmanaged;
+    public static void CopyTo<T>(this Buffer<T> buffer, T[] destination, int sourceOffset, int destinationOffset, int count) where T : unmanaged;
     public static void CopyTo<T>(this Buffer<T> buffer, Span<T> destination) where T : unmanaged;
     public static void CopyTo<T>(this Buffer<T> buffer, Span<T> destination, int offset) where T : unmanaged;
+    public static void CopyTo<T>(this Buffer<T> buffer, Buffer<T> destination) where T : unmanaged;
+    public static void CopyTo<T>(this Buffer<T> buffer, Buffer<T> destination, int sourceOffset, int destinationOffset, int count) where T : unmanaged;
     public static void CopyFrom<T>(this Buffer<T> buffer, T[] source) where T : unmanaged;
-    public static void CopyFrom<T>(this Buffer<T> buffer, T[] source, int sourceOffset, int bufferOffset, int count) where T : unmanaged;
+    public static void CopyFrom<T>(this Buffer<T> buffer, T[] source, int sourceOffset, int destinationOffset, int count) where T : unmanaged;
     public static void CopyFrom<T>(this Buffer<T> buffer, ReadOnlySpan<T> source) where T : unmanaged;        
     public static void CopyFrom<T>(this Buffer<T> buffer, ReadOnlySpan<T> source, int offset) where T : unmanaged;
+    public static void CopyFrom<T>(this Buffer<T> buffer, Buffer<T> source) where T : unmanaged;
+    public static void CopyFrom<T>(this Buffer<T> buffer, Buffer<T> source, int sourceOffset, int destinationOffset, int count) where T : unmanaged;
 }
 
 public static class StructuredBufferExtensions
 {
     public static void CopyTo<T>(this StructuredBuffer<T> buffer, ReadBackBuffer<T> destination) where T : unmanaged;
-    public static void CopyTo<T>(this StructuredBuffer<T> buffer, ReadBackBuffer<T> destination, int destinationOffset, int bufferOffset, int count) where T : unmanaged;
+    public static void CopyTo<T>(this StructuredBuffer<T> buffer, ReadBackBuffer<T> destination, int sourceOffset, int destinationOffset, int count) where T : unmanaged;
+    public static void CopyFrom<T>(this StructuredBuffer<T> buffer, UploadBuffer<T> source) where T : unmanaged;
+    public static void CopyFrom<T>(this StructuredBuffer<T> buffer, UploadBuffer<T> source, int sourceOffset, int destinationOffset, int count) where T : unmanaged;
 }

 public static class UploadBufferExtensions
 {
     public static void CopyTo<T>(this UploadBuffer<T> buffer, StructuredBuffer<T> destination) where T : unmanaged;
-    public static void CopyTo<T>(this UploadBuffer<T> buffer, StructuredBuffer<T> destination, int destinationOffset, int bufferOffset, int count) where T : unmanaged;
+    public static void CopyTo<T>(this UploadBuffer<T> buffer, StructuredBuffer<T> destination, int sourceOffset, int destinationOffset, int count) where T : unmanaged;
 }

+public static class ReadBackBufferExtensions
+{
+    public static void CopyFrom<T>(this ReadBackBuffer<T> buffer, StructuredBuffer<T> source) where T : unmanaged;
+    public static void CopyFrom<T>(this ReadBackBuffer<T> buffer, StructuredBuffer<T> source, int sourceOffset, int destinationOffset, int count) where T : unmanaged;
+}

 public static class Texture2DExtensions
 {
     public static T[,] ToArray<T>(this Texture2D<T> texture) where T : unmanaged;
+    public static T[,] ToArray<T>(this Texture2D<T> texture, int x, int y, int width, int height) where T : unmanaged;
     public static void CopyTo<T>(this Texture2D<T> texture, T[,] destination) where T : unmanaged;
     public static void CopyTo<T>(this Texture2D<T> texture, T[] destination) where T : unmanaged;
     public static void CopyTo<T>(this Texture2D<T> texture, T[] destination, Range x, Range y) where T : unmanaged;
     public static void CopyTo<T>(this Texture2D<T> texture, T[] destination, int x, int y, int width, int height) where T : unmanaged;
     public static void CopyTo<T>(this Texture2D<T> texture, T[] destination, int offset) where T : unmanaged;
     public static void CopyTo<T>(this Texture2D<T> texture, T[] destination, int offset, Range x, Range y) where T : unmanaged;
     public static void CopyTo<T>(this Texture2D<T> texture, T[] destination, int offset, int x, int y, int width, int height) where T : unmanaged;
     public static void CopyTo<T>(this Texture2D<T> texture, Span<T> destination) where T : unmanaged;
     public static void CopyTo<T>(this Texture2D<T> texture, Span<T> destination, Range x, Range y) where T : unmanaged;
     public static void CopyTo<T>(this Texture2D<T> texture, Span<T> destination, int x, int y, int width, int height) where T : unmanaged;
+    public static void CopyTo<T>(this Texture2D<T> texture, Texture2D<T> destination) where T : unmanaged;
+    public static void CopyTo<T>(this Texture2D<T> texture, Texture2D<T> destination, int sourceX, int sourceY, int width, int height) where T : unmanaged;
+    public static void CopyTo<T>(this Texture2D<T> texture, Texture2D<T> destination, int sourceX, int sourceY, int destinationX, int destinationY, int width, int height) where T : unmanaged;
     public static void CopyTo<T>(this Texture2D<T> texture, ReadBackTexture2D<T> destination) where T : unmanaged;
-    public static void CopyTo<T>(this Texture2D<T> texture, ReadBackTexture2D<T> destination, int x, int y, int width, int height) where T : unmanaged;
+    public static void CopyTo<T>(this Texture2D<T> texture, ReadBackTexture2D<T> destination, int sourceX, int sourceY, int width, int height) where T : unmanaged;
+    public static void CopyTo<T>(this Texture2D<T> texture, ReadBackTexture2D<T> destination, int sourceX, int sourceY, int destinationX, int destinationY, int width, int height) where T : unmanaged;
     public static void CopyFrom<T>(this Texture2D<T> texture, T[,] source) where T : unmanaged;
     public static void CopyFrom<T>(this Texture2D<T> texture, T[] source) where T : unmanaged;
     public static void CopyFrom<T>(this Texture2D<T> texture, T[] source, Range x, Range y) where T : unmanaged;
     public static void CopyFrom<T>(this Texture2D<T> texture, T[] source, int x, int y, int width, int height) where T : unmanaged;
     public static void CopyFrom<T>(this Texture2D<T> texture, T[] source, int offset, Range x, Range y) where T : unmanaged;
     public static void CopyFrom<T>(this Texture2D<T> texture, T[] source, int offset, int x, int y, int width, int height) where T : unmanaged;
     public static void CopyFrom<T>(this Texture2D<T> texture, ReadOnlySpan<T> source) where T : unmanaged;
     public static void CopyFrom<T>(this Texture2D<T> texture, ReadOnlySpan<T> source, Range x, Range y) where T : unmanaged;
     public static void CopyFrom<T>(this Texture2D<T> texture, ReadOnlySpan<T> source, int x, int y, int width, int height) where T : unmanaged;
+    public static void CopyFrom<T>(this Texture2D<T> texture, Texture2D<T> source) where T : unmanaged;
+    public static void CopyFrom<T>(this Texture2D<T> texture, Texture2D<T> source, int sourceX, int sourceY, int width, int height) where T : unmanaged;
+    public static void CopyFrom<T>(this Texture2D<T> texture, Texture2D<T> source, int sourceX, int sourceY, int destinationX, int destinationY, int width, int height) where T : unmanaged;
+    public static void CopyFrom<T>(this Texture2D<T> texture, UploadTexture2D<T> source) where T : unmanaged;
+    public static void CopyFrom<T>(this Texture2D<T> texture, UploadTexture2D<T> source, int sourceX, int sourceY, int width, int height) where T : unmanaged;
+    public static void CopyFrom<T>(this Texture2D<T> texture, UploadTexture2D<T> source, int sourceX, int sourceY, int destinationX, int destinationY, int width, int height) where T : unmanaged;
     public static void Save<T>(this Texture2D<T> texture, ReadOnlySpan<char> filename) where T : unmanaged;
 }

 public static class UploadTexture2DExtensions
 {
     public static void CopyTo<T>(this UploadTexture2D<T> texture, Texture2D<T> destination) where T : unmanaged;
-    public static void CopyTo<T>(this UploadTexture2D<T> texture, Texture2D<T> destination, int x, int y, int width, int height) where T : unmanaged;
+    public static void CopyTo<T>(this UploadTexture2D<T> texture, Texture2D<T> destination, int sourceX, int sourceY, int width, int height) where T : unmanaged;
+    public static void CopyTo<T>(this UploadTexture2D<T> texture, Texture2D<T> destination, int sourceX, int sourceY, int destinationX, int destinationY, int width, int height) where T : unmanaged;
 }

 public static class ReadBackTexture2DExtensions
 {
+    public static void CopyFrom<T>(this ReadBackTexture2D<T> texture, Texture2D<T> source) where T : unmanaged;
+    public static void CopyFrom<T>(this ReadBackTexture2D<T> texture, Texture2D<T> source, int sourceX, int sourceY, int width, int height) where T : unmanaged;
+    public static void CopyFrom<T>(this ReadBackTexture2D<T> texture, Texture2D<T> source, int sourceX, int sourceY, int destinationX, int destinationY, int width, int height) where T : unmanaged;
 }
 
 public static class Texture3DExtensions
 {
     public static T[,,] ToArray<T>(this Texture3D<T> texture) where T : unmanaged;
+    public static T[,,] ToArray<T>(this Texture3D<T> texture, int x, int y, int z, int width, int height, int depth) where T : unmanaged;
     public static void CopyTo<T>(this Texture3D<T> texture, T[,,] destination) where T : unmanaged;
     public static void CopyTo<T>(this Texture3D<T> texture, T[] destination, int offset) where T : unmanaged;
     public static void CopyTo<T>(this Texture3D<T> texture, T[] destination, int offset, Range x, Range y, Range z) where T : unmanaged;
     public static void CopyTo<T>(this Texture3D<T> texture, T[] destination, int offset, int x, int y, int z, int width, int height, int depth) where T : unmanaged;
     public static void CopyTo<T>(this Texture3D<T> texture, Span<T> destination) where T : unmanaged;
     public static void CopyTo<T>(this Texture3D<T> texture, Span<T> destination, Range x, Range y, Range z) where T : unmanaged;
     public static void CopyTo<T>(this Texture3D<T> texture, Span<T> destination, int x, int y, int z, int width, int height, int depth) where T : unmanaged;
+    public static void CopyTo<T>(this Texture3D<T> texture, Texture3D<T> destination) where T : unmanaged;
+    public static void CopyTo<T>(this Texture3D<T> texture, Texture3D<T> destination, int sourceX, int sourceY, int sourceZ, int width, int height, int depth) where T : unmanaged;
+    public static void CopyTo<T>(this Texture3D<T> texture, Texture3D<T> destination, int sourceX, int sourceY, int sourceZ, int destinationX, int destinationY, int destinationZ, int width, int height, int depth) where T : unmanaged;
     public static void CopyTo<T>(this Texture3D<T> texture, ReadBackTexture3D<T> destination) where T : unmanaged;
-    public static void CopyTo<T>(this Texture3D<T> texture, ReadBackTexture3D<T> destination, int x, int y, int z, int width, int height, int depth) where T : unmanaged;
+    public static void CopyTo<T>(this Texture3D<T> texture, ReadBackTexture3D<T> destination, int sourceX, int sourceY, int sourceZ, int width, int height, int depth) where T : unmanaged;
+    public static void CopyTo<T>(this Texture3D<T> texture, ReadBackTexture3D<T> destination, int sourceX, int sourceY, int sourceZ, int destinationX, int destinationY, int destinationZ, int width, int height, int depth) where T : unmanaged;
     public static void CopyFrom<T>(this Texture3D<T> texture, T[,,] source) where T : unmanaged;
     public static void CopyFrom<T>(this Texture3D<T> texture, T[] source) where T : unmanaged;
     public static void CopyFrom<T>(this Texture3D<T> texture, T[] source, Range x, Range y, Range z) where T : unmanaged;
     public static void CopyFrom<T>(this Texture3D<T> texture, T[] source, int x, int y, int z, int width, int height, int depth) where T : unmanaged;
     public static void CopyFrom<T>(this Texture3D<T> texture, T[] source, int offset, Range x, Range y, Range z) where T : unmanaged;
     public static void CopyFrom<T>(this Texture3D<T> texture, T[] source, int offset, int x, int y, int z, int width, int height, int depth) where T : unmanaged;
     public static void CopyFrom<T>(this Texture3D<T> texture, ReadOnlySpan<T> source) where T : unmanaged;
     public static void CopyFrom<T>(this Texture3D<T> texture, ReadOnlySpan<T> source, Range x, Range y, Range z) where T : unmanaged;
     public static void CopyFrom<T>(this Texture3D<T> texture, ReadOnlySpan<T> source, int x, int y, int z, int width, int height, int depth) where T : unmanaged;
+    public static void CopyFrom<T>(this Texture3D<T> texture, Texture3D<T> source) where T : unmanaged;
+    public static void CopyFrom<T>(this Texture3D<T> texture, Texture3D<T> source, int sourceX, int sourceY, int width, int height, int depth) where T : unmanaged;
+    public static void CopyFrom<T>(this Texture3D<T> texture, Texture3D<T> source, int sourceX, int sourceY, int sourceZ, int destinationX, int destinationY, int destinationZ, int width, int height) where T : unmanaged;
+    public static void CopyFrom<T>(this Texture3D<T> texture, UploadTexture3D<T> source) where T : unmanaged;
+    public static void CopyFrom<T>(this Texture3D<T> texture, UploadTexture3D<T> source, int sourceX, int sourceY, int sourceZ, int width, int height, int depth) where T : unmanaged;
+    public static void CopyFrom<T>(this Texture3D<T> texture, UploadTexture3D<T> source, int sourceX, int sourceY, int sourceZ, int destinationX, int destinationY, int destinationZ, int width, int height, int depth) where T : unmanaged;
 }

 public static class UploadTexture3DExtensions
 {
     public static void CopyTo<T>(this UploadTexture3D<T> texture, Texture3D<T> destination) where T : unmanaged;
-    public static void CopyTo<T>(this UploadTexture3D<T> texture, Texture3D<T> destination, int x, int y, int z, int width, int height, int depth) where T : unmanaged;
+    public static void CopyTo<T>(this UploadTexture3D<T> texture, Texture3D<T> destination, int sourceX, int sourceY, int sourceZ, int width, int height, int depth) where T : unmanaged;
+    public static void CopyTo<T>(this UploadTexture3D<T> texture, Texture3D<T> destination, int sourceX, int sourceY, int sourceZ, int destinationX, int destinationY, int destinationZ, int width, int height, int depth) where T : unmanaged;
 }

+public static class ReadBackTexture3DExtensions
+{
+    public static void CopyFrom<T>(this ReadBackTexture3D<T> texture, Texture3D<T> source) where T : unmanaged;
+    public static void CopyFrom<T>(this ReadBackTexture3D<T> texture, Texture3D<T> source, int sourceX, int sourceY, int sourceZ, int width, int height, int depth) where T : unmanaged;
+    public static void CopyFrom<T>(this ReadBackTexture3D<T> texture, Texture3D<T> source, int sourceX, int sourceY, int sourceZ, int destinationX, int destinationY, int destinationZ, int width, int height, int depth) where T : unmanaged;
+}
```

### Drawbacks

It's a breaking change for the current preview package, but given it's preview, not an actual issue.
It's also a bit of work 🙈

### Alternatives

Not really any, some things are just not possible without these APIs.

### PR DESCRIPTION
**Pull request by Sergio0694**
## Closes #158 

This PR includes additions and improvements to all resource copying APIs, as detailed in the linked issue.

cc. @hawkerm for review 😄

### CODE REVIEW

File: src/ComputeSharp/Graphics/Extensions/ReadBackTexture2DExtensions.cs [hunk 1]
```diff
@@ -8,6 +9,52 @@ namespace ComputeSharp;
...
+    /// <param name="sourceOffsetX">The horizontal offset in the source texture.</param>
+    /// <param name="sourceOffsetY">The vertical offset in the source texture.</param>
+    /// <param name="destinationOffsetX">The horizontal offset in the destination texture.</param>
+    /// <param name="destinationOffsetY">The vertical offset in the destination texture.</param>
+    /// <param name="width">The width of the memory area to copy.</param>
+    /// <param name="height">The height of the memory area to copy.</param>
+    public static void CopyFrom<T>(this ReadBackTexture2D<T> destination, Texture2D<T> source, int sourceOffsetX, int sourceOffsetY, int destinationOffsetX, int destinationOffsetY, int width, int height)
+        where T : unmanaged
+    {
+        source.CopyTo(destination, sourceOffsetX, sourceOffsetY, destinationOffsetX, destinationOffsetY, width, height);
```
Comment (hawkerm) [inline src/ComputeSharp/Graphics/Extensions/ReadBackTexture2DExtensions.cs:55]: Assuming the base function does all the bounds checking?

File: src/ComputeSharp/Graphics/Extensions/Texture2DExtensions.cs [hunk 2]
```diff
@@ -33,311 +60,467 @@ public static class Texture2DExtensions
...
+    /// <param name="sourceOffsetY">The vertical offset in the source texture.</param>
+    /// <param name="destinationOffsetX">The horizontal offset in the destination texture.</param>
+    /// <param name="destinationOffsetY">The vertical offset in the destination texture.</param>
     /// <param name="width">The width of the memory area to copy.</param>
     /// <param name="height">The height of the memory area to copy.</param>
-    public static void CopyTo<T>(this Texture2D<T> texture, ReadBackTexture2D<T> destination, int x, int y, int width, int height)
+    public static void CopyTo<T>(this Texture2D<T> source, ReadBackTexture2D<T> destination, int sourceOffsetX, int sourceOffsetY, int destinationOffsetX, int destinationOffsetY, int width, int height)
         where T : unmanaged
     {
-        texture.CopyTo(destination, 0, 0, x, y, width, height);
```
Comment (hawkerm) [inline src/ComputeSharp/Graphics/Extensions/Texture2DExtensions.cs:205]: Was this a bug before or am I mis-mapping from old to new above?

File: src/ComputeSharp/Graphics/Resources/Abstract/Texture2D{T}.cs [hunk 3]
```diff
@@ -223,17 +223,85 @@ internal void CopyTo(ref T destination, int size, int x, int y, int width, int h
...
+        destination.ThrowIfDisposed();
+
+        Guard.IsInRange(sourceOffsetX, 0, Width, nameof(sourceOffsetX));
+        Guard.IsInRange(sourceOffsetY, 0, Height, nameof(sourceOffsetY));
+        Guard.IsInRange(destinationOffsetX, 0, destination.Width, nameof(destinationOffsetX));
+        Guard.IsInRange(destinationOffsetY, 0, destination.Height, nameof(destinationOffsetY));
+        Guard.IsBetweenOrEqualTo(width, 1, Width, nameof(width));
+        Guard.IsBetweenOrEqualTo(height, 1, Height, nameof(height));
+        Guard.IsBetweenOrEqualTo(width, 1, destination.Width, nameof(width));
+        Guard.IsBetweenOrEqualTo(height, 1, destination.Height, nameof(height));
```
Comment (hawkerm) [inline src/ComputeSharp/Graphics/Resources/Abstract/Texture2D{T}.cs:252]: Are these needed? Aren't these a subset of the next cases of offset+width/height (as negative offsets aren't allowed)?

File: src/ComputeSharp/Graphics/Resources/Abstract/Texture2D{T}.cs [hunk 4]
```diff
@@ -147,22 +147,22 @@ private protected Texture2D(GraphicsDevice device, int width, int height, Resour
...
         GraphicsDevice.ThrowIfDisposed();
 
         ThrowIfDisposed();
 
-        Guard.IsInRange(x, 0, Width, nameof(x));
-        Guard.IsInRange(y, 0, Height, nameof(y));
+        Guard.IsInRange(sourceOffsetX, 0, Width, nameof(sourceOffsetX));
+        Guard.IsInRange(sourceOffsetY, 0, Height, nameof(sourceOffsetY));
         Guard.IsBetweenOrEqualTo(width, 1, Width, nameof(width));
         Guard.IsBetweenOrEqualTo(height, 1, Height, nameof(height));
```
Comment (hawkerm) [inline src/ComputeSharp/Graphics/Resources/Abstract/Texture2D{T}.cs:163]: Aren't these checked by the checking the offset is positive and offset+size is in bounds check below?

File: src/ComputeSharp/Graphics/Resources/Abstract/Texture2D{T}.cs [hunk 5]
```diff
@@ -242,18 +310,18 @@ internal void CopyTo(ReadBackTexture2D<T> destination, int destinationX, int des
...
-        Guard.IsInRange(sourceX, 0, Width, nameof(sourceX));
-        Guard.IsInRange(sourceY, 0, Height, nameof(sourceY));
+        Guard.IsInRange(sourceOffsetX, 0, Width, nameof(sourceOffsetX));
+        Guard.IsInRange(sourceOffsetY, 0, Height, nameof(sourceOffsetY));
+        Guard.IsInRange(destinationOffsetX, 0, destination.Width, nameof(destinationOffsetX));
+        Guard.IsInRange(destinationOffsetY, 0, destination.Height, nameof(destinationOffsetY));
         Guard.IsBetweenOrEqualTo(width, 1, Width, nameof(width));
         Guard.IsBetweenOrEqualTo(height, 1, Height, nameof(height));
         Guard.IsBetweenOrEqualTo(width, 1, destination.Width, nameof(width));
         Guard.IsBetweenOrEqualTo(height, 1, destination.Height, nameof(height));
```
Comment (hawkerm) [inline src/ComputeSharp/Graphics/Resources/Abstract/Texture2D{T}.cs:320]: Same here and elsewhere?

File: src/ComputeSharp/Graphics/Resources/Debug/BufferDebugView{T}.cs [hunk 6]
```diff
@@ -20,7 +20,7 @@ public BufferDebugView(Buffer<T>? buffer)
         {
             var items = GC.AllocateUninitializedArray<T>(buffer.Length);
 
-            buffer.CopyTo(ref items[0], buffer.Length, 0);
+            buffer.CopyTo(ref items[0], 0, buffer.Length);
```
Comment (hawkerm) [inline src/ComputeSharp/Graphics/Resources/Debug/BufferDebugView{T}.cs:23]: Were these the types of breaking changes you were talking about? Bit confused on why the order needed to change here? (Probably as looking at code on Github is hard... 😋)

File: tests/ComputeSharp.Tests/Attributes/AdditionalResourceAttribute.cs [hunk 7]
```diff
@@ -0,0 +1,17 @@
+﻿using System;
+
+namespace ComputeSharp.Tests.Attributes;
+
+/// <summary>
+/// An attribute to use with <see cref="CombinatorialAttribute"/> targeting an additional specific resource type.
+/// </summary>
+[AttributeUsage(AttributeTargets.Method, AllowMultiple = true)]
+public sealed class AdditionalResourceAttribute : Attribute
```
Comment (hawkerm) [inline tests/ComputeSharp.Tests/Attributes/AdditionalResourceAttribute.cs:9]: What does this do? I'm a bit confused on what this sets up.

**Review by hawkerm — COMMENTED** [review]
Going to see if I can test out with the sample I was building, went though with some questions/comments though for you. Thanks! 😊

**Comment by hawkerm** [comment]
These were the weird failing tests right? Known to fail sometimes locally on certain machines?

image

image

**Comment by hawkerm** [comment]
Still had to modify the `IShaderRunner` interface, but think that was expected still:
```cs
public void Execute(ReadWriteTexture2D<Rgba32, Float4> texture, TimeSpan timespan);
```

File: src/ComputeSharp/Graphics/Extensions/ReadBackTexture2DExtensions.cs (hunk 1 above)
Comment (Sergio0694) [reply src/ComputeSharp/Graphics/Extensions/ReadBackTexture2DExtensions.cs:55]: That is correct, this calls `Texture2D<T>.CopyTo`, which does all the paarameter validation:

https://github.com/Sergio0694/ComputeSharp/blob/da90dd9157aaa07ca6834a50380e65ce1914a0f3/src/ComputeSharp/Graphics/Resources/Abstract/Texture2D%7BT%7D.cs#L313-L324

File: src/ComputeSharp/Graphics/Extensions/Texture2DExtensions.cs (hunk 2 above)
Comment (Sergio0694) [reply src/ComputeSharp/Graphics/Extensions/Texture2DExtensions.cs:205]: This is just the parameter being renamed, plus the swapped order.
We're now always doing, in order: source, destination, count.
Should make the APIs less error prone across overloads as they're more consistent 🙂

File: src/ComputeSharp/Graphics/Resources/Abstract/Texture2D{T}.cs (hunk 3 above)
Comment (Sergio0694) [reply src/ComputeSharp/Graphics/Resources/Abstract/Texture2D{T}.cs:252]: They're needed, yeah. You could have (in fact, I accidentally had in a test, which is how I spotted this ahah) a valid offset that's some positive number, and then a negative length that's small enough. If you don't check the length explicitly, you could end up with a final offset + length being valid, but with the length still being a negative number 🙂

File: src/ComputeSharp/Graphics/Resources/Abstract/Texture2D{T}.cs (hunk 4 above)
Comment (Sergio0694) [reply src/ComputeSharp/Graphics/Resources/Abstract/Texture2D{T}.cs:163]: See answer above.

File: src/ComputeSharp/Graphics/Resources/Abstract/Texture2D{T}.cs (hunk 5 above)
Comment (Sergio0694) [reply src/ComputeSharp/Graphics/Resources/Abstract/Texture2D{T}.cs:320]: Same as above 🙂

File: src/ComputeSharp/Graphics/Resources/Debug/BufferDebugView{T}.cs (hunk 6 above)
Comment (Sergio0694) [reply src/ComputeSharp/Graphics/Resources/Debug/BufferDebugView{T}.cs:23]: Yup 😄
The reason for the change is the same I mentioned above: it allows all APIs to be consistent now across different overloads. Now regardless of whatever method you're calling you _always_ have source parameters first, then destination parameters, then count, both when copying to and when copying from. Makes it much harder to accidentally swap the source and destination parameters when doing copies.

File: tests/ComputeSharp.Tests/Attributes/AdditionalResourceAttribute.cs (hunk 7 above)
Comment (Sergio0694) [reply tests/ComputeSharp.Tests/Attributes/AdditionalResourceAttribute.cs:9]: Ahah this is a magic attribute for my `[CombinatorialTestMethod]` setup.
You can see it being used here:

https://github.com/Sergio0694/ComputeSharp/blob/da90dd9157aaa07ca6834a50380e65ce1914a0f3/tests/ComputeSharp.Tests/BufferTests.cs#L227-L240

What this does is it adds another combinatorial sequence of optional arguments to a given test method. For instance, in this case this method is actually being invoked a total of `[DEVICES]x[RESOURCES]x[ADDITIONAL_RESOURCES]x[DATA]` times. That is, this single method gets invoked a total of 90 times with all possible combinations of parameters specified here, without the need to write all those tests manually 😄

**Comment by Sergio0694** [comment]
@hawkerm That is correct, those tests are somewhat bugged, but only on some machines and only under very specific cases. I'll see if I can get a repro together this weeked so that I can ping the D3D12MA engineer as well as the WARP team if needed 😄

And yes, changing that input texture is something we can look into in another PR.
What was your use case again for needing to copy it?

**Comment by hawkerm** [comment]
> And yes, changing that input texture is something we can look into in another PR. What was your use case again for needing to copy it?

For #137 to be able to grab a copy of the result texture to pass in as input to the next run of the shader. As the texture passed in is empty each loop through right?

**Comment by Sergio0694** [comment]
> "For #137 to be able to grab a copy of the result texture to pass in as input to the next run of the shader."

Ah, right. Yeah in that case you'd need a temporary buffer every frame otherwise you'd have shader invocations do side effect on the other pixels that are being read by other invocations (if every invocation is not just reading a single pixel at that corresponding location, but a window of pixels or any other group of pixels in general), and things would get messed up.

> "As the texture passed in is empty each loop through right?"

No, the input texture will always have the data from the previous frame. So if that's enough for you and you're just having each pixel read on the corresponding pixel from the previous frame, you don't really need the additional temporary buffer.

**Pull request merged**